### DISCUSSION TITLE
error building vips on Freebsd 13.2

### DISCUSSION THREAD

**Discussion by tbswork1** [libvips/libvips#3617] [General]
```
[212/499] Compiling C++ object libvips/resample/libresample.a.p/reduceh.cpp.o
FAILED: libvips/resample/libresample.a.p/reduceh.cpp.o
c++ -Ilibvips/resample/libresample.a.p -Ilibvips/resample -I../libvips/resample -I. -I.. -Ilibvips/include -I../libvips/include -Ilibvips/include/vips -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include -fvisibility=hidden -fcolor-diagnostics -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++11 -O3 -DG_DISABLE_CAST_CHECKS -DG_DISABLE_CHECKS -DG_DISABLE_ASSERT -Werror=pointer-arith -fPIC -DHAVE_CONFIG_H=1 -pthread -MD -MQ libvips/resample/libresample.a.p/reduceh.cpp.o -MF libvips/resample/libresample.a.p/reduceh.cpp.o.d -o libvips/resample/libresample.a.p/reduceh.cpp.o -c ../libvips/resample/reduceh.cpp
In file included from ../libvips/resample/reduceh.cpp:52:
In file included from /usr/local/include/glib-2.0/glib/gi18n-lib.h:23:
In file included from /usr/local/include/glib-2.0/glib.h:34:
In file included from /usr/local/include/glib-2.0/glib/gasyncqueue.h:34:
In file included from /usr/local/include/glib-2.0/glib/gthread.h:34:
In file included from /usr/local/include/glib-2.0/glib/gatomic.h:30:
In file included from /usr/local/include/glib-2.0/glib/glib-typeof.h:43:
In file included from /usr/include/c++/v1/type_traits:420:
/usr/include/c++/v1/cstddef:37:10: fatal error: 'version' file not found
#include <version>
         ^~~~~~~~~
1 error generated.
```

**Comment by lovell** [comment]
Which version of libvips? Which compiler and version?

**Reply by tbswork1** [reply]
@lovell 
Hi, I am using the latest version 8.14.4, and the compiler:

`FreeBSD clang version 14.0.5 `

May I know why you change to use meson as the build tool?

**Reply by jcupitt** [reply]
`meson` is pretty nice, and fixes a lot of longstanding issues with autotools. And autotools is becoming slowly less supported, so I think most projects are moving away from it now.

**Reply by jcupitt** [reply]
Oh, maybe you mean why not cmake?

I've used cmake a lot at the uni with itk, mirtk, and packages like that, and the scripting language drives me crazy, plus it's such a huge thing (800,000 lines of C++, really?!?!?!) which makes it difficult to deploy and manage.

Meson is quite a large improvement over cmake in terms of simplicity, updateability (it's just `pip install meson`) and speed, in my opinion.

All build systems have issues of course, and I certainly wouldn't claim meson was flawless. It's a bit like vi vs. emacs.

**Comment by tbswork1** [comment]
I copied the missing header files and now the build went through. I get the following test result:

```
> meson test
ninja: no work to do.
ninja: Entering directory `git/libvips/build'
ninja: no work to do.
1/9 formats                 FAIL             0.02s   exit status 1
>>> MALLOC_PERTURB_=108 git/libvips/test/test_formats.sh

2/9 seq                     FAIL             0.02s   exit status 1
>>> MALLOC_PERTURB_=155 git/libvips/test/test_seq.sh

3/9 stall                   SKIP             0.01s   exit status 77
4/9 cli                     FAIL             0.07s   exit status 1
>>> MALLOC_PERTURB_=106 git/libvips/test/test_cli.sh

5/9 connections             SKIP             0.01s   exit status 77
6/9 descriptors             SKIP             0.03s   exit status 77
7/9 webpsave_timeout        SKIP             0.01s   exit status 77
8/9 fuzz                    OK               0.59s
9/9 threading               FAIL             0.71s   exit status 1
>>> MALLOC_PERTURB_=115 /git/libvips/test/test_threading.sh


Ok:                 1
Expected Fail:      0
Fail:               4

```

**Comment by jcupitt** [comment]
I installed freebsd 13.2 in virtualbox and built libvips git master and it all seems to work well. All the tests pass too:

![image](https://github.com/libvips/libvips/assets/580843/409dc9e6-4e55-47c1-af94-8a2ea500918b)

I did `pkg install vips` to pull in most of the dependencies for me, then just:

```
git clone https://github.com/libvips/libvips.git
cd libvips
meson setup build --prefix /home/john/vips
cd build
ninja
ninja install
ninja test
```

This is with clang 14.0.5 (the standard one). I think you need to give more information before we can reproduce this.

**Comment by tbswork1** [comment]
@jcupitt 
Thanks. I use your sequence and it produces error free test result. Don't know why the meson commands in README.md gives the problem.

Now there is errors linking my program with libvips:

```
[  1%] Linking CXX executable tbssvc
ld: error: undefined symbol: vips_image_new
>>> referenced by thumbnailutil.c:260 (/home/user/mylibs/thumbnailutil.c:260)
>>>               CMakeFiles/tbssvc.dir/home/user/mylibs/thumbnailutil.c.o:(thumbnail_process)
```

I use the following in CMakefile.list

```
set(ENV{PKG_CONFIG_PATH} "$ENV{PKG_CONFIG_PATH}:/home/user/vips/lib/pkgconfig")
find_package(PkgConfig REQUIRED)                            # for using pkg_search_module()
pkg_search_module(PKG_VIPS REQUIRED IMPORTED_TARGET vips)   # this looks for vips.pc file

message(STATUS "PKG_VIPS_INCLUDE_DIRS: '${PKG_VIPS_INCLUDE_DIRS}'")
```

the above debug output should include `/home/user/vips/include` but it doesn't. This works on Linux though. I therefore have to hardcode the vips path in order to compile.

**Reply by jcupitt** [reply]
I tried:

```
$ export PKG_CONFIG_PATH=/home/vips/vips/libdata/pkgconfig
$ pkg-config vips --modversion
8.15.0
```

I guess freebsd uses `libdata` for `pkgconfig`, which is a bit odd.

After that I could run:

```
$ cd libvips/examples
$ cc new-from-buffer.c `pkg-config vips --cflags --libs`
```

And it worked ok.

**Comment by tbswork1** [comment]
I run the above command and get: 

```
cc new-from-buffer.c `pkg-config vips --cflags --libs`
cc: error: unsupported option '--cflags'
cc: error: unsupported option '--libs`'
cc: error: no such file or directory: '`pkg-config'
cc: error: no such file or directory: 'vips'

```

**Reply by jcupitt** [reply]
> cc: error: no such file or directory: '`pkg-config'
> cc: error: unsupported option '--libs`'

That's strange, are you using a shell that doesn't support backticks? Try with `sh`.

**Reply by tbswork1** [reply]
Yes I am using fish. Thanks.

**Reply by jcupitt** [reply]
Then you'll need to use sh, or find the equivalent of backticks in fish.
